Delete non-terminal jobs and subworkflow invocations when cancelling invocation

mvdbeek opened the pull request:
Came up as a wish during the last UI/UX call with @nomadscientist. I contemplated making the job deletion part optional, but couldn't think of a reason why we wouldn't make this the only default. No problem to make it configurable though if there is a reason to keep the old default.

Not sure this is completely free of race conditions, but worst case you gotta delete again ? That seems fine.

I think it'd also be nice to have various flavors of bulk deletion (delete all datasets produced by an invocation, delete all non-output datasets, maybe other modalities ?).

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek committed:
Delete non-terminal jobs and subworkflow invocation when cancelling an invocation

nsoranzo commented on lib/galaxy/managers/workflows.py:
```diff
@@ -413,9 +413,21 @@ def cancel_invocation(self, trans, decoded_invocation_id):
+
+        for step in workflow_invocation.steps:
+            for job in step.jobs:
+                job.mark_deleted()
```
For consistency with `galaxy.managers.jobs.JobManager.stop()`:

```suggestion
                job.mark_deleted(self.app.config.track_jobs_in_database)
```

nsoranzo replied:
Actually, it would probably be better to use `galaxy.managers.jobs.JobManager.stop()`:

```python
from galaxy.managers.jobs import JobManager

...
job_manager = JobManager(trans.app)
...

                job_manager.stop(flush=False)
```

with this additional patch:

```diff
diff --git a/lib/galaxy/managers/jobs.py b/lib/galaxy/managers/jobs.py
index 54c01bcea4..18589609c8 100644
--- a/lib/galaxy/managers/jobs.py
+++ b/lib/galaxy/managers/jobs.py
@@ -239,12 +239,13 @@ class JobManager:
         trans.sa_session.refresh(job)
         return job
 
-    def stop(self, job, message=None):
+    def stop(self, job, message=None, flush=True):
         if not job.finished:
             job.mark_deleted(self.app.config.track_jobs_in_database)
-            session = self.app.model.session
-            with transaction(session):
-                session.commit()
+            if flush:
+                session = self.app.model.session
+                with transaction(session):
+                    session.commit()
             self.app.job_manager.stop(job, message=message)
             return True
         else:
diff --git a/lib/galaxy/structured_app.py b/lib/galaxy/structured_app.py
index 22f03c10e1..c83718a2d7 100644
--- a/lib/galaxy/structured_app.py
+++ b/lib/galaxy/structured_app.py
@@ -41,6 +41,7 @@ from galaxy.workflow.trs_proxy import TrsProxy
 if TYPE_CHECKING:
     from galaxy.config_watchers import ConfigWatchers
     from galaxy.jobs import JobConfiguration
+    from galaxy.jobs.manager import JobManager
     from galaxy.managers.collections import DatasetCollectionManager
     from galaxy.managers.hdas import HDAManager
     from galaxy.managers.histories import HistoryManager
@@ -109,7 +110,7 @@ class MinimalManagerApp(MinimalApp):
     installed_repository_manager: "InstalledRepositoryManager"
     user_manager: Any
     job_config: "JobConfiguration"
-    job_manager: Any  # galaxy.jobs.manager.JobManager
+    job_manager: "JobManager"
     job_metrics: JobMetrics
     dynamic_tool_manager: Any  # 'galaxy.managers.tools.DynamicToolManager'
     genomes: "Genomes"

```

BTW, it is very confusing that we have 2 `JobManager` classes and `app.job_manager` is not an instance of `galaxy.managers.jobs.JobManager` !

mvdbeek replied:
I think the conditional flush in the diff is mandatory since the JobHandlerStopQueue receives the job by id, and I believe the queue itself uses a different thread and should have a different sqlalchemy session that doesn't see the changes to the job state yet.

My understanding of how this currently works is that the deleted state gets picked up by the monitor without needing to call `self.app.job_manager.stop(job, message=message)`. (And maybe existing code should be refactored to not call `app.job_manager.stop` ?)

nsoranzo commented on lib/galaxy/managers/workflows.py:
```diff
@@ -413,9 +413,21 @@ def cancel_invocation(self, trans, decoded_invocation_id):
+                trans.sa_session.add(job)
+            if step.implicit_collection_jobs:
+                for icjja in step.implicit_collection_jobs.jobs:
+                    icjja.job.mark_deleted()
```
```suggestion
                    icjja.job.mark_deleted(self.app.config.track_jobs_in_database)
```

mvdbeek replied:
I ... would maybe open a different PR and remove the option that has been deprecated for some years ?

nsoranzo replied:
Sure, but that PR needs to be merged before this one if you don't make this change since the default in the method is False:
```
    def mark_deleted(self, track_jobs_in_database=False):
```
and in the config file is True.

mvdbeek replied:
I don't think it matters, the handling for deleted_new and deleted is the same?

mvdbeek committed:
Display cancel button while invocation jobs not terminal

mvdbeek committed:
Replace for loop with sql update statement

mvdbeek committed:
Make job terminal check in JobWrapper.change_state more robust

By always fetching the job state from the database.

mvdbeek committed:
More WIP stuff

mvdbeek committed:
Make state change logic atomic

mvdbeek committed:
Make invocation cancellation two-step process

API / users set invocation to cancelling, scheduler then deletes
outputs. This avoids race conditions where the cancelled invocation
still generates new jobs.

mvdbeek committed:
Refactor ItemGrabber into JobGrabber and InvocationGrabber

mvdbeek committed:
Fix job state history

mvdbeek committed:
Fix invocation cancel test

mvdbeek committed:
Fix class name logging

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -1370,7 +1371,7 @@ class Job(Base, JobLike, UsesCreateAndUpdateTime, Dictifiable, Serializable):
     states = JobState
 
-    terminal_states = [states.OK, states.ERROR, states.DELETED]
+    terminal_states = [states.OK, states.ERROR, states.DELETED, states.DELETING]
```
I find confusing to include `DELETING` here, since I'd expect that the job state changes to DELETED at a certain point, so DELETING is not terminal.

mvdbeek replied:
I can rename it to almost_terminal_states 😅 I'll see what the alternatives are

nsoranzo replied:
Could be added explicitly when you are checking for (almost) terminal states?

mvdbeek replied:
Moved that the finished_states, which we already use in the `finished` property

nsoranzo replied:
Perfect!

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -1640,10 +1641,30 @@ def all_entry_points_configured(self):
     def set_state(self, state):
         """
-        Save state history
+        Save state history. Returns True if stat has changed, else False
```
```suggestion
        Save state history. Returns True if state has changed, else False
```

nsoranzo commented on lib/galaxy/workflow/scheduling_manager.py:
```diff
@@ -332,6 +331,12 @@ def __attempt_schedule(self, invocation_id, workflow_scheduler):
             try:
+                if workflow_invocation.state == workflow_invocation.states.CANCELLING:
+                    workflow_invocation.cancel_invocation_steps()
+                    workflow_invocation.state = workflow_invocation.states.CANCELLED
```
```suggestion
                    workflow_invocation.mark_cancelled()
```

Otherwise `mark_cancelled()` seems unused.

mvdbeek committed:
Apply Nicola's suggestions

mvdbeek committed:
Add new finished_states and move deleting to that

nsoranzo commented on lib/galaxy/model/__init__.py:
```diff
@@ -1634,12 +1632,32 @@ def all_entry_points_configured(self):
+            # generate statement that will not revert DELETING or DELETED back to anything non-terminal
+            rval = session.execute(
+                update(Job.table)
+                .where(Job.table.c.id == self.id, ~Job.table.c.state.in_((Job.states.DELETING, Job.states.DELETED)))
```
I think we are moving away from using `.table.c` for SA 2.0 , ping @jdavcs

mvdbeek replied:
Thanks, updated that and the code I touched

jdavcs replied:
Yes: if `Foo` is a mapped class (not a subquery, etc.) and `bar` is a mapped attribute, then `Foo.table.c.bar` is equivalent to `Foo.bar`. So for the sake of simplicity, `Foo.bar`.

nsoranzo commented on lib/galaxy/jobs/handler.py:
```diff
@@ -240,9 +261,8 @@ def __init__(self, app: MinimalManagerApp, dispatcher):
             self.app.job_config.handler_assignment_methods
```
Doesn't make a difference, but maybe swap `ItemGrabber` with `JobGrabber` on the line above this?

nsoranzo commented on lib/galaxy_test/api/test_workflows.py:
```diff
@@ -4034,12 +4034,80 @@ def test_cancel_workflow_invocation(self):
             message = invocation["messages"][0]
             assert message["reason"] == "user_request"
 
+    @skip_without_tool("collection_creates_dynamic_nested")
```
```suggestion
    @skip_without_tool("cat_data_and_sleep")
    @skip_without_tool("identifier_multiple")
```

mvdbeek committed:
Use sqlalchemy 2.0 query syntax

mvdbeek committed:
Use JobGrabber.get_grabbable_handler_assignment_method static method for consistency

mvdbeek committed:
Prevent cancelled invocation resetting back to new or ready
